Implement struct enumerator for Tensor<T>

NewellClark opened the pull request:
This fixes [issue 28391](https://github.com/dotnet/runtime/issues/28391).

NewellClark committed:
Merge remote-tracking branch 'dotnet/master'

NewellClark committed:
Merge remote-tracking branch 'dotnet/master'

NewellClark committed:
Merge remote-tracking branch 'dotnet/master'

NewellClark committed:
Merge remote-tracking branch 'dotnet/master'

NewellClark committed:
Added struct enumerator to Tensor<T>

Fix issue #28391

NewellClark committed:
Added struct enumerator to Tensor<T>

Added unit test for element enumeration.

Fix issue #28391

NewellClark committed:
Merge remote-tracking branch 'origin/enhancement-issue-28391' into enhancement-issue-28391

NewellClark committed:
Added xml doc comments

NewellClark commented:
Oh dear. The bot applied the wrong tag. It should be `area-System.Numerics.Tensors`.

gfoidl commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,56 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+        /// </summary>
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly Tensor<T> tensor;
+            private int index;
```
Nit: naming according the guidelines
```suggestion
            private readonly Tensor<T> _tensor;
            private int _index;
```

SingleAccretion replied:
Hmm, I believe the idea here is to follow the file's naming conventions:
https://github.com/dotnet/runtime/blob/aa327625efca011f8f96cbd59f002fd1b28d2f56/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs#L258

NewellClark replied:
Yeah that was my thought process. I initially named with underscores but changed it to be consistent with the rest of the file.

gfoidl commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,56 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+            {
+                Debug.Assert(tensor != null);
+
+                this.tensor = tensor;
```
```suggestion
                _tensor = tensor;
```
See https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md

gfoidl commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,56 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+
+            void IEnumerator.Reset()
+            {
+                index = 0;
+                Current = default;
```
Could be
```suggestion
                this = default;
```
too.

NewellClark replied:
I did not know you could do that. I'll update it.

NewellClark committed:
Apply suggested edit to System.Numerics.Tensor.Enumerator

Co-authored-by: Günther Foidl <gue@korporal.at>

NewellClark reviewed (commented):
Not sure why github is asking me to review my own pull request. I left the names the same to be consistent with the rest of the file, but applied the `this = default` suggested edit.

NewellClark commented:
Okay, the runtime CI check failed with a "503 Service Not Available" result. I'm guessing this wasn't caused by the suggested changes. Gonna quickly close and reopen to try running them again.

NewellClark commented:
Now one of the tests seems to have crashed. Been running for ~6 hours.

eiriktsarpalis commented:
@eerhardt @danmosemsft do we need to put this through API review?

eerhardt commented:
> do we need to put this through API review?

This library hasn't been officially approved yet, nor has a stable version of the package been shipped. So my opinion is that we can iterate on this prototype library without bringing each change to API review. And then when we are happy with the whole shape, we bring the whole library to API review at once.

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,55 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+                }
+            }
+
+            object? IEnumerator.Current => Current;
```
This shouldn't be needed.  If removing it results in errors, I expect the problem is the use of the auto-prop on Current including a private set; it'd be better to instead just write out the field manually and remove the setter from the property:
```C#
private T _current;
...
public T Current => _current;
```

NewellClark replied:
Hmm it still gives me an error when I remove the explicit implementation of `Current`, even when I remove the auto-prop and create the backing field explicitly. 
`Error: CS0738	'Tensor<T>.Enumerator' does not implement interface member 'IEnumerator.Current'. 'Tensor<T>.Enumerator.Current' cannot implement 'IEnumerator.Current' because it does not have the matching return type of 'object'.	System.Numerics.Tensors (net461), System.Numerics.Tensors (netstandard1.1), System.Numerics.Tensors (netstandard2.0)`

stephentoub replied:
Ah, right.  Ignore me: hadn't had my coffee yet :)

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,55 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+        /// </summary>
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly Tensor<T> tensor;
```
_tensor

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,55 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly Tensor<T> tensor;
+            private int index;
```
_index

NewellClark committed:
Fixed naming

NewellClark committed:
Forgot to change back Current to auto-prop

NewellClark commented:
Build just failed due to flaky test (issue #32805). Rerunning...

jeffhandley commented:
@stephentoub I believe this is ready for you re-review when you have time.

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,55 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+                this = default;
+            }
+
+            void IDisposable.Dispose() { }
```
Generally we make the Dispose implementation public.

NewellClark replied:
Fixed.

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/Tensor.cs:
```diff
@@ -691,6 +691,55 @@ private Tensor<T> GetTriangle(int offset, bool upper)
+
+            void IEnumerator.Reset()
+            {
+                this = default;
```
This is going to overwrite _tensor and make subsequent use fail.  If the goal is to not implement it, it should throw a NotSupportedException.  Otherwise, it should reset _index and Current but leave _tensor alone.

NewellClark replied:
Yikes, not sure how I didn't notice that blatant null-reference bug. Fixed.

stephentoub commented on src/libraries/System.Numerics.Tensors/tests/TensorTests.cs:
```diff
@@ -2419,5 +2419,18 @@ public void TestIReadOnlyTMembers(TensorConstructor constructor)
+            }
+
+            Assert.Equal(GetExpected(tensor), tensor);
+        }
```
Other things generally worth testing are disposing the enumerator multiple times, and reseting and using the enumerator again if reset is implemented.

NewellClark replied:
I've added tests for idempotent dispose and proper reset behavior. 

Is it okay for dispose to be a no-op, and for the enumerator to not throw when used after being disposed? I noticed most other struct enumerators in dotnet/runtime are implemented that way, I assume for performance.

stephentoub replied:
> Is it okay for dispose to be a no-op, and for the enumerator to not throw when used after being disposed? 

Yes.  You can see, for example, that the enumerators generated by the C# compiler often nop Dispose:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4oDchhGAzOVqQMKkDehp35jFS5HPQA8ASwB2AFwB8pAKLiArgFsYUAIaSYANXUAbRTADOACgCUXHpwI9bpAGbRSJiZNKjSAXlLEq70sKkOL7uANShZoI45ADs7jQ2PAC+hElAA==

NewellClark committed:
Apply suggested changes

-Make Reset and Dispose public
-Fix null ref bug with Reset
-Added tests for Reset and Dispose

NewellClark committed:
Merge remote-tracking branch 'dotnet/master' into enhancement-issue-28391
